articles/zk-snarks-under-the-hood/zk_snarks_under_the_hood.md

```diff
@@ -0,0 +1,147 @@
+---
+Title: "Zk-SNARKs: Under the Hood"
+Status: Done
+Level: 6
+Note: There is a few places that are hard to fully understand
+---
+
+# Introduction
+
+Remember that the solution to a QAP is a set of polynomials (A, B, C) such that A(x) * B(x) - C(x) = H(x) * Z(x), where:
+
+- A is a linear combination of a set of polynomials {A_1…A_m}
+- B is the linear combination of {B_1…B_m} with the same coefficients
+- C is a linear combination of {C_1…C_m} with the same coefficients 
+
+Coefficients are $1, x, x^2, x^3, ... , x^m$
+
+Note that from this step, we are going to use [elliptic curve](../../terms/elliptic_curve.md), so we will have a generator point G (corresponding to 1 in the elliptic curve) appear in our equation as a way to convert a normal number into an elliptic curve.
+
+However, in a large system with thousands of logic gates like a hash function or something similar, the list that the prover needs to send to the verifier would be too long. So, we use [knowledge of exponent](../../terms/knowledge_of_exponent.md) assumption to make it shorter.
+
+On the other hand, using the prefixed function Z = $(x-1)(x-2)(x-3)(x-4)$ as a divider for the prover's QAP is not secure enough as it would create room for the prover to make "fake" proofs.
+
+Here, we use random secret variables $t, k_a, k_b, k_c$ generated by the verifier, then encrypt them and send those encrypted numbers to the prover. t will be a point where we evaluate our polynomial. $t, k_a, k_b$, and $k_c$ are “toxic waste,” numbers that absolutely must be deleted at all costs, or else whoever has them will be able to make fake proofs.
+
+Hence, what actually happens is that we add the following values to the [trusted setup](../../terms/trusted_setup.md):
+
+- $G * A_1(t), G * A_1(t) * k_a$
+- $G * A_2(t), G * A_2(t) * k_a$
+- …
+- $G * B_1(t), G * B_1(t) * k_b$
+- $G * B_2(t), G * B_2(t) * k_b$
+- …
+- $G * C_1(t), G * C_1(t) * k_c$
+- $G * C_2(t), G * C_2(t) * k_c$
+- …
+
+Now, if someone gives you a pair of points P, Q such that $P * k_a = Q$ (reminder: we don’t need $k_a$ to check this, as we can do a [pairing](../../terms/elliptic_curve.md#elliptic-curve-pairings) check), then you know that what they are giving you is a linear combination of $A_i$ polynomials evaluated at t. By evaluating polynomials at t, we reduce the need to send the whole matrix of the polynomials; however, this raises a need to check for polynomials' coefficients in case the prover colluded and a need to check for $k_a, k_b, k_c$ in case the prover inserts $k_a', k_b', k_c'$ which $P' * k_a' = Q$ into the proof later on.
+
+For example, instead of sending $a*G + b*G*t + c*G*t^2$, the prover will send a point $a'*G$ instead.
+
+Our coefficients in the linear combination at the end of the QAP step were $1, x, x^2, x^3, ... , x^m$ but we are going to evaluate our polynomials at t, so they will turn into $G, G*t, G*t^2, G*t^3, ..., G*t^m$.
```

Comment:
It's because we want to "map" from 1,x, x^2 to points in elliptic curve. Not related to t

Reply:
x here is a unknown while t is a variable, we can't find an "unknown" point though